protected.go

```diff
@@ -58,26 +56,9 @@ func (addr *protectedAddr) TCPAddr() *net.TCPAddr {
 	return &net.TCPAddr{IP: addr.IP, Port: addr.Port}
 }
 
-// New construct a protector from the protect function and DNS server IP address.
-func New(protect Protect, dnsServerIP string) *Protector {
-	ipAddr := parseIP(dnsServerIP)
-	if ipAddr == nil {
-		log.Debugf("Invalid DNS server IP %s, default to %s", dnsServerIP, defaultDNSServer)
-		ipAddr = parseIP(defaultDNSServer)
-	}
-
-	dnsAddrString := fmt.Sprintf("%v:%d", ipAddr, dnsPort)
-	ipv4Addr := ipAddr.To4()
-	if ipv4Addr != nil {
-		log.Debug("Using IPv4 DNS server")
-		dnsAddr := syscall.SockaddrInet4{Port: dnsPort}
-		copy(dnsAddr.Addr[:], ipv4Addr)
-		return &Protector{protect, &dnsAddr, dnsAddrString, false}
-	}
-	log.Debug("Using IPv6 DNS server")
-	dnsAddr := syscall.SockaddrInet6{Port: dnsPort}
-	copy(dnsAddr.Addr[:], ipAddr)
-	return &Protector{protect, &dnsAddr, dnsAddrString, true}
+// New construct a protector from the protect function and function that provides a DNS server IP address.
+func New(protect Protect, dnsServerIP func() string) *Protector {
```

Comment:
I changed dnsServerIP to a function so that we can always get the latest DNS server IP. That way, if we switch from a V4 to a V6 network or back, we can respond appropriately in here.